ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/InsertValuesExecutor.java

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.rest.server.execution;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import io.confluent.ksql.GenericRow;
+import io.confluent.ksql.KsqlExecutionContext;
+import io.confluent.ksql.logging.processing.NoopProcessingLogContext;
+import io.confluent.ksql.metastore.model.StructuredDataSource;
+import io.confluent.ksql.parser.tree.AstVisitor;
+import io.confluent.ksql.parser.tree.Expression;
+import io.confluent.ksql.parser.tree.InsertValues;
+import io.confluent.ksql.parser.tree.Literal;
+import io.confluent.ksql.parser.tree.Node;
+import io.confluent.ksql.parser.tree.NullLiteral;
+import io.confluent.ksql.rest.entity.KsqlEntity;
+import io.confluent.ksql.services.ServiceContext;
+import io.confluent.ksql.statement.ConfiguredStatement;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.KsqlException;
+import io.confluent.ksql.util.SchemaUtil;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.LongSupplier;
+import java.util.stream.Collectors;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.connect.data.Field;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.Schema.Type;
+
+public class InsertValuesExecutor {
```

Comment:
It feels to me that the logic for doing the production of messages to Kafka should be in the engine, not the rest-api.  This executor should call the engine with the command.

This will also mean we get headless for free...

Reply:
This can be done in a follow-up PR, but there isn't a good place as of today to put it in the engine that I know of. This definitely isn't an `ExecutableDdlStatement` and it doesn't create a persistent query. Furthermore, I don't want this to be persisted onto the command topic and after #2712 there won't be any unreplicated statements that the engine executes, so this would be another edge case we need to handle.

> This will also mean we get headless for free...

 I know we disagree on this, but this is something that I actively don't want in V1 so for me that's not a reason to make this change. When we want to move this to headless mode, we need to give it more thought and if we move it now we can't take it away.

Reply:
Hummm... I still feel this should be in the engine! However, happy for this to be a follow up task as long as its not ignored.